hwy/contrib/unroller/unroller-inl.h

```diff
@@ -0,0 +1,496 @@
+// Copyright 2023 Matthew Kolbe
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if defined(HIGHWAY_HWY_CONTRIB_UNROLLER_UNROLLER_INL_H_) == \
+    defined(HWY_TARGET_TOGGLE)
+#ifdef HIGHWAY_HWY_CONTRIB_UNROLLER_UNROLLER_INL_H_
+#undef HIGHWAY_HWY_CONTRIB_UNROLLER_UNROLLER_INL_H_
+#else
+#define HIGHWAY_HWY_CONTRIB_UNROLLER_UNROLLER_INL_H_
+#endif
+
+#include <cstring>
+#include <limits>
+
+#include "hwy/highway.h"
+
+HWY_BEFORE_NAMESPACE();
+namespace hwy {
+namespace HWY_NAMESPACE {
+
+namespace hn = hwy::HWY_NAMESPACE;
+
+template <class DERIVED, typename IN_T, typename OUT_T>
+struct UnrollerUnit {
+  DERIVED* me() { return static_cast<DERIVED*>(this); }
+
+  static constexpr inline size_t UnitLanes() {
+    return HWY_MIN(HWY_MAX_LANES_D(hn::ScalableTag<IN_T>),
+                   HWY_MAX_LANES_D(hn::ScalableTag<OUT_T>));
+  }
+
+  using IT = hn::CappedTag<IN_T, UnitLanes()>;
+  using OT = hn::CappedTag<OUT_T, UnitLanes()>;
+  IT d_in;
+  OT d_out;
+
+  inline hn::Vec<OT> Func(const ptrdiff_t idx,
+                          hn::Vec<IT> const& HWY_RESTRICT x,
+                          hn::Vec<OT> const& HWY_RESTRICT y) {
+    return me()->Func(idx, x, y);
+  }
+
+  inline hn::Vec<IT> X0Init() { return me()->X0InitImpl(); }
+
+  inline hn::Vec<IT> X0InitImpl() { return hn::Zero(d_in); }
+
+  inline hn::Vec<OT> YInit() { return me()->YInitImpl(); }
+
+  inline hn::Vec<OT> YInitImpl() { return hn::Zero(d_out); }
+
+  inline hn::Vec<IT> Load(ptrdiff_t const& HWY_RESTRICT idx,
```

Comment:
Probably easier and faster to pass word-sized variables by value, without const&.